### PR TITLE
Fix1736

### PR DESCRIPTION
**Pull request by IwanKaramazow**
Fixes https://github.com/facebook/reason/issues/1736

This PR adds printing of braces around JSX as JSX-arguments:
```reason
<Description term={<Text text="Age" />}> ... </Description>;
<Foo bar={<Baz />} />
```
Error messages for ambiguous cases are also provided:
captura de pantalla 2018-01-12 a las 21 35 02
captura de pantalla 2018-01-12 a las 21 35 57

### CODE REVIEW

**Comment by ubsan** [comment]
This seems unfortunate. Could we make it context dependent instead?

**Comment by jordwalke** [comment]
Great! So, if we convert some code from OCaml to Reason that uses `/>>`, how is it printed? I think we need a way to express  the `/>>`identifier with some test cases showing that it is possible.

**Comment by IwanKaramazow** [comment]
Per discussion with @chenglou, braces are printed around JSX as jsx-arguments:
```reason
<Description term={<Text text="Age" />}> ... </Description>;
<Foo bar={<Baz />} />
```
The infix operators are preserved this way, also added tests, and it's a bit more familiar for JS-developers.
Error messages for ambiguous cases are also provided:
captura de pantalla 2018-01-12 a las 21 35 02
captura de pantalla 2018-01-12 a las 21 35 57

**Comment by thangngoc89** [comment]
Why don’t we use normal () braces? Those curly braces look ugly

**Comment by chenglou** [comment]
@thangngoc89 mostly for familiarity with existing JSX, and to reduce the amount of parentheses.

**Comment by jordwalke** [comment]
To reduce the number of parens, we could print them as `{ }` in any other JSX cases that require parens.

File: src/reason-parser/reason_pprint_ast.ml [hunk 1]
```diff
@@ -2452,6 +2452,26 @@ let is_direct_pattern x = x.ppat_attributes = [] && match x.ppat_desc with
   | Ppat_construct ( {txt= Lident"()"}, None) -> true
   | _ -> false
 
+let isJSXComponent loc args =
```
Comment (jordwalke) [inline src/reason-parser/reason_pprint_ast.ml]: Probably worth pulling out JSX logic into its own module in the future.

**Pull request merged**